desktop/ui/src/main/java/org/datacleaner/panels/SchemaTreePanel.java

```diff
@@ -39,73 +48,147 @@
 
 public class SchemaTreePanel extends DCPanel {
 
-	private static final long serialVersionUID = 1L;
-
-	private final InjectorBuilder _injectorBuilder;
-	private JComponent _updatePanel;
-
-	@Inject
-	protected SchemaTreePanel(InjectorBuilder injectorBuilder) {
-		super(WidgetUtils.COLOR_DEFAULT_BACKGROUND);
-		_injectorBuilder = injectorBuilder;
-		setLayout(new BorderLayout());
-		setBorder(new EmptyBorder(4, 4, 4, 4));
-		setDatastore(null, false);
-	}
-
-	public void setDatastore(final Datastore datastore, final boolean expandTree) {
-		removeAll();
-		if (datastore == null) {
-			add(new DCPanel().setPreferredSize(150, 150), BorderLayout.CENTER);
-			return;
-		}
-
-		add(new LoadingIcon().setPreferredSize(150, 150), BorderLayout.CENTER);
-
-		// load the schema tree in the background because it will retrieve
-		// metadata about the datastore (might take several seconds)
-		new SwingWorker<SchemaTree, Void>() {
-			@Override
-			protected SchemaTree doInBackground() throws Exception {
-				Injector injector = _injectorBuilder.with(Datastore.class, datastore).createInjector();
-				SchemaTree tree = injector.getInstance(SchemaTree.class);
-				return tree;
-			}
-
-			protected void done() {
-				try {
-					SchemaTree schemaTree = get();
-					final JScrollPane schemaTreeScroll = WidgetUtils.scrolleable(schemaTree);
-					schemaTreeScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
-					schemaTree.addComponentListener(new ComponentAdapter() {
-						@Override
-						public void componentResized(ComponentEvent e) {
-							updateParentPanel();
-						}
-
-					});
-					removeAll();
-					add(schemaTreeScroll, BorderLayout.CENTER);
-					updateParentPanel();
+    private static final long serialVersionUID = 1L;
+
+    private static final String DEFAULT_SEARCH_FIELD_TEXT = "Search component library...";
+
+    private final InjectorBuilder _injectorBuilder;
+    private JComponent _updatePanel;
+    private final JTextField _searchTextField;
```

Comment:
One thing I notice that could go wrong here .... The searchTextField is a single instance for this panel. But the schema tree seems to be loaded for every time setDatastore(...) is called. That means the same search textbox might be used on multiple trees. And will thus be instrumented many times by multiple swing workers. That doesn't seem to work very well in this code. I think the instantiation of the textbox should follow the same lifecycle as the instantiation/injection of the schema tree.